### PR TITLE
[Python] Add documentation of auto-instrumenting Python services using the operator

### PR DESCRIPTION
**Pull request by TylerHelmuth**
Adds a new section for Python automatic docs detailing how to use the OpenTelemetry operator to inject auto-instrumentation.

If we like this I can do something similar for the other languages the operator can inject.

### CODE REVIEW

File: content/en/docs/instrumentation/python/automatic/operator.md
Comment (chalin) [inline content/en/docs/instrumentation/python/automatic/operator.md:90]: old --> older

File: content/en/docs/instrumentation/python/automatic/operator.md
Comment (cartermp) [inline content/en/docs/instrumentation/python/automatic/operator.md:8]:
```suggestion
If you run your Python service in Kubernetes, you can take advantage of the [OpenTelemetry Operator](https://github.com/open-telemetry/opentelemetry-operator) to inject auto-instrumentation without having to modify each of your services directly.
```

File: content/en/docs/instrumentation/python/automatic/operator.md
Comment (cartermp) [inline content/en/docs/instrumentation/python/automatic/operator.md:12]:
```suggestion
First, install the [OpenTelemetry Operator](https://github.com/open-telemetry/opentelemetry-operator) into your cluster.
```

File: content/en/docs/instrumentation/python/automatic/operator.md
Comment (cartermp) [inline content/en/docs/instrumentation/python/automatic/operator.md:14]:
```suggestion
You can do this with the [Operator release manifest](https://github.com/open-telemetry/opentelemetry-operator#getting-started), the [Operator helm chart](https://github.com/open-telemetry/opentelemetry-helm-charts/tree/main/charts/opentelemetry-operator#opentelemetry-operator-helm-chart), or with [Operator Hub](https://operatorhub.io/operator/opentelemetry-operator).  
```

File: content/en/docs/instrumentation/python/automatic/operator.md
Comment (cartermp) [inline content/en/docs/instrumentation/python/automatic/operator.md:16]:
```suggestion
In most cases, you will need to install [cert-manager](https://cert-manager.io/docs/installation/). If you use the helm chart, there is an option to generate a self-signed cert instead.
```

File: content/en/docs/instrumentation/python/automatic/operator.md
Comment (cartermp) [inline content/en/docs/instrumentation/python/automatic/operator.md:20]:
```suggestion
It is a best practice to send telemetry from containers to an [OpenTelemetry Collector](../../../../collector/) instead of directly to a backend.  The Collector helps simplify secret management, decouples data export problems (such as a need to do retries) from your apps, and lets you add additional data to your telemetry, such as with the `k8sattributesprocessor` component. If you chose not to use a Collector, you can skip to the next section.
```

File: content/en/docs/instrumentation/python/automatic/operator.md
Comment (cartermp) [inline content/en/docs/instrumentation/python/automatic/operator.md:22]:
```suggestion
The Operator provides a [Custom Resource Definition (CRD) for the OpenTelemetry Collector](https://github.com/open-telemetry/opentelemetry-operator/blob/main/docs/api.md#opentelemetrycollector) which is used to create an instance of the Collector that the Operator manages. The following example deploys the Collector as a deployment (the default), but there are other [deployment modes](https://github.com/open-telemetry/opentelemetry-operator#deployment-modes) that can be used.
```

File: content/en/docs/instrumentation/python/automatic/operator.md
Comment (cartermp) [inline content/en/docs/instrumentation/python/automatic/operator.md:61]:
```suggestion
The above command results in a deployment of the Collector that you can use as an endpoint for auto-instrumentation in your pods.
```

File: content/en/docs/instrumentation/python/automatic/operator.md
Comment (cartermp) [inline content/en/docs/instrumentation/python/automatic/operator.md:63]:
```suggestion
## Configure Autoinstrumentation
```

File: content/en/docs/instrumentation/python/automatic/operator.md
Comment (cartermp) [inline content/en/docs/instrumentation/python/automatic/operator.md:65]:
```suggestion
To be able to manage autoinstrumentation, the Operator needs to be configured to know what pods to instrument and which autoinstrumentation to use for those pods. This is done via the [Instrumentation CRD](https://github.com/open-telemetry/opentelemetry-operator/blob/main/docs/api.md#instrumentation).
```

File: content/en/docs/instrumentation/python/automatic/operator.md
Comment (cartermp) [inline content/en/docs/instrumentation/python/automatic/operator.md:67]: I don't think I understand how this section relates to the code sample below. Where does it specify that we're using the Python SDK?

File: content/en/docs/instrumentation/python/automatic/operator.md
Comment (cartermp) [inline content/en/docs/instrumentation/python/automatic/operator.md:78]: Where did `demo-collector` come from? I see below that you wrote:

> which is the Collector's otlpreceiver's `http` port.

So this is the default name of the endpoint when you add the collector CRD?

File: content/en/docs/instrumentation/python/automatic/operator.md
Comment (cartermp) [inline content/en/docs/instrumentation/python/automatic/operator.md:88]:
```suggestion
By default, the Instrumentation resource that auto-instruments python services uses the `otlp` with the `http/proto` protocol. This means that the configured endpoint must be able to receive OTLP over `http/proto`.  Therefore, the example uses `http://demo-collector:4318`, which is the Collector's otlpreceiver's `http` port.
```

File: content/en/docs/instrumentation/python/automatic/operator.md
Comment (cartermp) [inline content/en/docs/instrumentation/python/automatic/operator.md:90]:
```suggestion
> As of operator v0.67.0, the `OTEL_EXPORTER_OTLP_TRACES_PROTOCOL` and `OTEL_EXPORTER_OTLP_METRICS_PROTOCOL` env vars are automatically set to `http/proto`.  If you use an older version of the Operator you **MUST** set these env variables to `http/proto`, or python auto-instrumentation will not work.
```

File: content/en/docs/instrumentation/python/automatic/operator.md
Comment (cartermp) [inline content/en/docs/instrumentation/python/automatic/operator.md:92]:
```suggestion
```

Let's remove this for now -- we can add more stuff for other languages later (I'll create a tracking issue).

File: content/en/docs/instrumentation/python/automatic/operator.md
Comment (cartermp) [inline content/en/docs/instrumentation/python/automatic/operator.md:94]:
```suggestion
Now that your Instrumentation object is created, your cluster has the ability to auto-instrument python services and send data to an endpoint. However, auto-instrumentation with the OpenTelemetry Operator follows an opt-in model. In order to activate autoinstrumentation, you'll need to add an annotation to your deployment.
```

File: content/en/docs/instrumentation/python/automatic/operator.md
Comment (cartermp) [inline content/en/docs/instrumentation/python/automatic/operator.md:121]:
```suggestion

```

File: content/en/docs/instrumentation/python/automatic/operator.md
Comment (cartermp) [inline content/en/docs/instrumentation/python/automatic/operator.md:98]:
```suggestion
The final step is to opt in your python services to autoinstrumentation.  This is done by updating your service’s `spec.template.metadata.annotations` to include `instrumentation.opentelemetry.io/inject-python: "true"`.

Alternatively, the annotation can be added to a namespace, which will result in all services in that namespace to opt-in to autoinstrumentation. See the [Operators auto-instrumentation documentation](https://github.com/open-telemetry/opentelemetry-operator/blob/main/README.md#opentelemetry-auto-instrumentation-injection) for more details.
```

**Review by cartermp — CHANGES_REQUESTED** [review]
Round of copy-edits and some questions. Really happy to see this come together.

File: content/en/docs/instrumentation/python/automatic/operator.md
Comment (TylerHelmuth) [reply content/en/docs/instrumentation/python/automatic/operator.md:67]: This paragraph is supposed to be a warning that configuring the Instrumentation resource is both the trickiest but most important step of injecting auto-instrumentation.  I was trying to highlight that some of the items set in the Instrumentation resource map to items in the SDK that the operator configures and then inject.  

When I was learning this process I got held up for a second realizing that setting all the env vars in the Instrumentation object was the same as setting all the env var for an application that implemented the SDK itself.  In order to correctly configure the Instrumentation resource you have to understand the env vars that the SDK uses/depend on.

How does this rewording sound?

```suggestion
Creating the Instrumentation resource correctly is paramount to getting auto-instrumentation working.  Making sure all endpoints and env vars are correct is required for auto-instrumentation to work properly.
```

File: content/en/docs/instrumentation/python/automatic/operator.md
Comment (TylerHelmuth) [reply content/en/docs/instrumentation/python/automatic/operator.md:78]: The OpenTelemetryCollector resourse automatically creates a service with the name of the OpenTelemetryCollector resource prepended to `-collector`.  Since our OpenTelemetryCollector resource was named `demo` we communicate with it as `demo-collector`.  I will call out that naming in the collector section.

Comment (TylerHelmuth) [reply content/en/docs/instrumentation/python/automatic/operator.md:80]: Updated

**Comment by svrnm** [comment]
thanks @TylerHelmuth 

I am wondering if this should live below https://opentelemetry.io/docs/k8s-operator/ ? Where do you think people might expect it?

cc @pavolloffay

**Comment by pavolloffay** [comment]
> I am wondering if this should live below https://opentelemetry.io/docs/k8s-operator/ ? Where do you think people might expect it?

I would prefer to keep it in the operator docs, but crosslink it from the python instrumentation docs.

**Review by pavolloffay — COMMENTED** [review]
Good work, 

as mentioned in another comment. A better place might be directly under operator docs.

We should also think how other operator auto-instrumentations will be documented and try to reuse common sections (collector deployment etc.)

**Comment by cartermp** [comment]
@pavolloffay Yes, agreed. I think we can iterate on this. We can extract various sections as subpages and share (e.g., collector stuff, some commentary) or break some of it up into separate articles (although only if it doesn't complicate the end-user experience).

@TylerHelmuth how would you like to proceed? We can take it in as-is and I can tweak from there, or we can try to make adjustments in this PR. I can PR against your branch if that works.

**Review by chalin — COMMENTED** [review]
Hi all. If someone could run the **markdown Prettifier** over the new file before it is merged, that would be great! Otherwise it can be handled later (esp. since @cartermp is considering iterative improvements).

**Comment by TylerHelmuth** [comment]
I can update this PR on Tuesday to reorganize the content.

File: content/en/docs/instrumentation/python/automatic/operator.md
Comment (cartermp) [reply content/en/docs/instrumentation/python/automatic/operator.md:69]: Yep, I like it.

File: content/en/docs/k8s-operator/automatic.md
Comment (cartermp) [inline content/en/docs/k8s-operator/automatic.md:75]:
```suggestion
### .NET
```

File: content/en/docs/k8s-operator/automatic.md
Comment (cartermp) [inline content/en/docs/k8s-operator/automatic.md:83]:
```suggestion
### Node.js
```

File: content/en/docs/k8s-operator/automatic.md
Comment (cartermp) [inline content/en/docs/k8s-operator/automatic.md:119]:
```suggestion
- .NET: `instrumentation.opentelemetry.io/inject-dotnet: "true"`
```

File: content/en/docs/k8s-operator/automatic.md
Comment (cartermp) [inline content/en/docs/k8s-operator/automatic.md:121]:
```suggestion
- Node.js: `instrumentation.opentelemetry.io/inject-nodejs: "true"`
```

File: content/en/docs/k8s-operator/automatic.md
Comment (cartermp) [inline content/en/docs/k8s-operator/automatic.md:14]:
```suggestion
## Installation
```

File: content/en/docs/k8s-operator/automatic.md
Comment (cartermp) [inline content/en/docs/k8s-operator/automatic.md:22]:
```suggestion
## Create an OpenTelemetry Collector (Optional)
```

File: content/en/docs/k8s-operator/automatic.md
Comment (cartermp) [inline content/en/docs/k8s-operator/automatic.md:69]:
```suggestion
## Configure Autoinstrumentation
```

File: content/en/docs/k8s-operator/automatic.md
Comment (cartermp) [inline content/en/docs/k8s-operator/automatic.md:79]:
```suggestion
### Java
```

File: content/en/docs/k8s-operator/automatic.md
Comment (cartermp) [inline content/en/docs/k8s-operator/automatic.md:87]:
```suggestion
### Python
```

File: content/en/docs/k8s-operator/automatic.md
Comment (cartermp) [inline content/en/docs/k8s-operator/automatic.md:115]:
```suggestion
## Add annotations to existing deployments
```

**Review by cartermp — COMMENTED** [review]
Some copy-edits. Mainly reducing header sizes, since there's a top-level walkthrough but all the content is in subsections, so we can probably just move those subsections up one level.

**Review by cartermp — APPROVED** [review]
This sparks joy.

@pavolloffay and @open-telemetry/python-approvers can you take a look please?

**Comment by TylerHelmuth** [comment]
@pavolloffay @svrnm @cartermp @chalin

Moved content under k8s-operator and tried to keep the "walkthrough" feel of the doc.  My thought was that the ins and outs of the different operator CRs and other configuration options is best handled in the operator's documentation and this section should be used as a "I want to auto-instrument a language, give me step by step instructions".

File: content/en/docs/instrumentation/python/automatic/operator.md
Comment (svrnm) [inline content/en/docs/instrumentation/python/automatic/operator.md:12]:
```suggestion
[See the OpenTelemetry Operator Auto-instrumentation docs for more details.](/docs/k8s-operator/automatic/)
```

File: content/en/docs/k8s-operator/automatic.md
Comment (svrnm) [inline content/en/docs/k8s-operator/automatic.md:11]:
```suggestion
auto-instrumentation libraries for .NET, Java, NodeJS and Python
```

File: content/en/docs/k8s-operator/automatic.md
Comment (svrnm) [inline content/en/docs/k8s-operator/automatic.md:36]:
```suggestion
to your telemetry, such as with the [k8sattributesprocessor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/k8sattributesprocessor) component. If you
```

File: content/en/docs/k8s-operator/automatic.md
Comment (svrnm) [inline content/en/docs/k8s-operator/automatic.md:117]: Can you put a sentence to introduce what you are going to do with the following command?

**Review by svrnm — CHANGES_REQUESTED** [review]

**Review by svrnm — APPROVED** [review]

**Comment by TylerHelmuth** [comment]
@svrnm each section of yaml is already ready to be copy and pasted into a terminal and use kubectl to apply the resource.  Is there more you think we should add?  

I purposefully added the ability to copy and paste the sections but didn't want to get into details about kubectl itself or the various methods of applying resource in k8s.  My thinking was that if the reader feels comfortable using an operator (which I view as an advanced k8s technique), then they should already know how to apply resources using kubectl.

**Comment by svrnm** [comment]
> @svrnm each section of yaml is already ready to be copy and pasted into a terminal and use kubectl to apply the resource. Is there more you think we should add?
> 
> I purposefully added the ability to copy and paste the sections but didn't want to get into details about kubectl itself or the various methods of applying resource in k8s. My thinking was that if the reader feels comfortable using an operator (which I view as an advanced k8s technique), then they should already know how to apply resources using kubectl.

I know plenty of people that use the operator without really knowing a lot of basics for k8s ;-) ... but I get your point. Giving one way of many and calling out that there are alternatives would be my preferred solution, but that's optional and from my point of view we should be good to get this into the docs?

@open-telemetry/operator-approvers any concerns?

**Review by srikanthccv — APPROVED** [review]
Thanks @TylerHelmuth

**Comment by cartermp** [comment]
@open-telemetry/operator-approvers -- any feedback? I'll merge within the week since it's been vetted by several folks at this point

**Review by pavolloffay — APPROVED** [review]

**Comment by svrnm** [comment]
🎉

**Pull request merged**
